src/main/kotlin/io/libp2p/pubsub/gossip/GossipScore.kt

```diff
@@ -0,0 +1,213 @@
+package io.libp2p.pubsub.gossip
+
+import io.libp2p.core.PeerId
+import io.libp2p.core.pubsub.ValidationResult
+import io.libp2p.etc.types.cappedDouble
+import io.libp2p.etc.types.createLRUMap
+import io.libp2p.etc.types.millis
+import io.libp2p.etc.util.P2PService
+import pubsub.pb.Rpc
+import java.util.Optional
+import java.util.concurrent.ScheduledExecutorService
+import java.util.concurrent.ScheduledFuture
+import java.util.concurrent.TimeUnit
+import kotlin.math.max
+import kotlin.math.min
+import kotlin.math.pow
+
+class GossipScore(
+    val params: GossipScoreParams = GossipScoreParams(),
+    val executor: ScheduledExecutorService,
+    val curTime: () -> Long
+) {
+
+    inner class TopicScores(val topic: Topic) {
+        private val params = topicParams[topic]
+
+        var joinedMeshTime: Long = 0
+        var firstMessageDeliveries: Double by cappedDouble(
+            0.0,
+            this@GossipScore.params.decayToZero,
+            params.FirstMessageDeliveriesCap
+        )
+        var meshMessageDeliveries: Double by cappedDouble(
+            0.0,
+            this@GossipScore.params.decayToZero,
+            params.MeshMessageDeliveriesCap
+        )
+        var meshFailurePenalty: Double by cappedDouble(0.0, this@GossipScore.params.decayToZero)
+        var invalidMessages: Double by cappedDouble(0.0, this@GossipScore.params.decayToZero)
+
+        fun inMesh() = joinedMeshTime > 0
+
+        fun meshTimeNorm() = min(
+            (if (inMesh()) curTime() - joinedMeshTime else 0).toDouble() / params.TimeInMeshQuantum.toMillis(),
+            params.TimeInMeshCap
+        )
+
+        fun isMeshMessageDeliveriesActive() =
+            inMesh() && ((curTime() - joinedMeshTime).millis > params.MeshMessageDeliveriesActivation)
```

Comment:
Spec again suggests this `curTime()` to be calculated lazily during the decay update.

Reply:
Commented above